internal/manager/setup.go

```diff
@@ -190,86 +192,65 @@ func setupAdmissionServer(
-	dataplaneAddressFinder := dataplane.NewAddressFinder()
-	udpDataplaneAddressFinder := dataplane.NewAddressFinder()
-	var getter func(ctx context.Context) ([]string, error)
-	if c.UpdateStatus {
-		// Default
-		if overrideAddrs := c.PublishStatusAddress; len(overrideAddrs) > 0 {
-			dataplaneAddressFinder.SetOverrides(overrideAddrs)
-		} else if c.PublishService.String() != "" {
-			publishServiceNn := c.PublishService
-			dataplaneAddressFinder.SetGetter(func(ctx context.Context) ([]string, error) {
-				svc := new(corev1.Service)
-				if err := mgrc.Get(ctx, publishServiceNn, svc); err != nil {
-					return nil, err
-				}
+func setupDataplaneAddressFinder(mgrc client.Client, c *Config, log logr.Logger) (*dataplane.AddressFinder, *dataplane.AddressFinder, error) {
+	if !c.UpdateStatus {
+		return nil, nil, nil
+	}
 
-				var addrs []string
-				switch svc.Spec.Type { //nolint:exhaustive
-				case corev1.ServiceTypeLoadBalancer:
-					for _, lbaddr := range svc.Status.LoadBalancer.Ingress {
-						if lbaddr.IP != "" {
-							addrs = append(addrs, lbaddr.IP)
-						}
-						if lbaddr.Hostname != "" {
-							addrs = append(addrs, lbaddr.Hostname)
-						}
-					}
-				default:
-					addrs = append(addrs, svc.Spec.ClusterIPs...)
-				}
+	defaultAddressFinder, err := buildDataplaneAddressFinder(mgrc, c.PublishStatusAddress, c.PublishService)
+	if err != nil {
+		return nil, nil, fmt.Errorf("status updates enabled but no method to determine data-plane addresses: %w", err)
+	}
+	udpAddressFinder, err := buildDataplaneAddressFinder(mgrc, c.PublishStatusAddressUDP, c.PublishServiceUDP)
+	if err != nil {
+		log.Info("falling back to a default address finder for UDP", "reason", err.Error())
+		udpAddressFinder = defaultAddressFinder
+	}
 
-				if len(addrs) == 0 {
-					return nil, fmt.Errorf("waiting for addresses to be provisioned for publish service %s", publishServiceNn)
-				}
+	return defaultAddressFinder, udpAddressFinder, nil
+}
 
-				return addrs, nil
-			})
-		} else {
-			return nil, nil, fmt.Errorf("status updates enabled but no method to determine data-plane addresses, need either --publish-service or --publish-status-address")
-		}
+func buildDataplaneAddressFinder(mgrc client.Client, publishStatusAddress []string, publishServiceNn types.NamespacedName) (*dataplane.AddressFinder, error) {
+	addressFinder := dataplane.NewAddressFinder()
 
-		// UDP. falls back to default if not configured
-		if udpOverrideAddrs := c.PublishStatusAddressUDP; len(udpOverrideAddrs) > 0 {
-			udpDataplaneAddressFinder.SetUDPOverrides(udpOverrideAddrs)
-		} else if c.PublishServiceUDP.String() != "" {
-			publishServiceNn := c.PublishServiceUDP
-			udpDataplaneAddressFinder.SetGetter(func(ctx context.Context) ([]string, error) {
-				svc := new(corev1.Service)
-				if err := mgrc.Get(ctx, publishServiceNn, svc); err != nil {
-					return nil, err
-				}
+	if overrideAddrs := publishStatusAddress; len(overrideAddrs) > 0 {
```

Comment:
Not sure why we need this variable if we already have the function arg 🤔

Reply:
Me neither 😅. It got here somehow, but it's definitely not needed: 18785e0